// astal-local/gtk4/EventBox.tsx
import Gtk from "gi://Gtk?version=4.0"
import GObject from "gi://GObject?version=2.0"
import Gdk from "gi://Gdk?version=4.0"
import astalify, { type ConstructProps } from "./astalify.js"

function filter(children: any[]) {
    return children.flat(Infinity).map(ch => ch instanceof Gtk.Widget
        ? ch
        : new Gtk.Label({ visible: true, label: String(ch) }))
}

class EventBoxClass extends Gtk.Box {
  static {
    GObject.registerClass({
      GTypeName: "EventBox",
      Signals: {
        "button-press-event": { param_types: [GObject.TYPE_OBJECT], return_type: GObject.TYPE_BOOLEAN },
        "button-release-event": { param_types: [GObject.TYPE_OBJECT], return_type: GObject.TYPE_BOOLEAN },
        "scroll-event": { param_types: [GObject.TYPE_OBJECT], return_type: GObject.TYPE_BOOLEAN },
        "enter-notify-event": { param_types: [GObject.TYPE_OBJECT], return_type: GObject.TYPE_BOOLEAN },
        "leave-notify-event": { param_types: [GObject.TYPE_OBJECT], return_type: GObject.TYPE_BOOLEAN },
        "motion-notify-event": { param_types: [GObject.TYPE_OBJECT], return_type: GObject.TYPE_BOOLEAN },
        "hover": { param_types: [], return_type: GObject.TYPE_BOOLEAN },
        "hover-lost": { param_types: [], return_type: GObject.TYPE_BOOLEAN },
      },
    }, this)
  }

  constructor() {
    super()
    const controller = new Gtk.EventControllerLegacy()
    controller.connect("event", (_: Gtk.EventController, event: Gdk.Event) => {
      const eventType = Gdk.Event.get_event_type(event)

      let signalName: string | null = null
      switch (eventType) {
        case Gdk.EventType.BUTTON_PRESS:
        case Gdk.EventType.DOUBLE_BUTTON_PRESS:
        case Gdk.EventType.TRIPLE_BUTTON_PRESS:
          signalName = "button-press-event"
          break
        case Gdk.EventType.BUTTON_RELEASE:
          signalName = "button-release-event"
          break
        case Gdk.EventType.SCROLL:
          signalName = "scroll-event"
          break
        case Gdk.EventType.ENTER_NOTIFY:
          this.emit("hover")
          signalName = "enter-notify-event"
          break
        case Gdk.EventType.LEAVE_NOTIFY:
          this.emit("hover-lost")
          signalName = "leave-notify-event"
          break
        case Gdk.EventType.MOTION_NOTIFY:
          signalName = "motion-notify-event"
          break
      }

      if (signalName) {
        return this.emit(signalName, this)
      }

      return false
    })
    this.add_controller(controller)
  }

  get_child() {
    return this.get_first_child()
  }
}

export type EventBoxProps = ConstructProps<EventBoxClass, Gtk.Box.ConstructorProps>

export const EventBox = astalify(EventBoxClass, {
    getChildren(self) {
        const children: Array<Gtk.Widget> = []
        let ch = self.get_first_child()
        while (ch !== null) {
            children.push(ch)
            ch = ch.get_next_sibling()
        }
        return children
    },
    setChildren(self, children) {
        for (const child of filter(children)) {
            self.append(child)
        }
    },
})
